pkg/autodiscovery/providers/docker.go

```diff
@@ -25,16 +26,22 @@ const (
 // DockerConfigProvider implements the ConfigProvider interface for the docker labels.
 type DockerConfigProvider struct {
 	sync.RWMutex
-	dockerUtil *docker.DockerUtil
-	upToDate   bool
-	streaming  bool
-	health     *health.Handle
+	dockerUtil   *docker.DockerUtil
+	upToDate     bool
+	streaming    bool
+	health       *health.Handle
+	labelCache   map[string]map[string]string
+	syncInterval int
+	syncCounter  int
 }
 
 // NewDockerConfigProvider returns a new ConfigProvider connected to docker.
 // Connectivity is not checked at this stage to allow for retries, Collect will do it.
 func NewDockerConfigProvider(config config.ConfigurationProviders) (ConfigProvider, error) {
-	return &DockerConfigProvider{}, nil
+	return &DockerConfigProvider{
+		// periodically resync if we're missing events
+		syncInterval: 30,
```

Comment:
can we make it clearer (either in the comment or in the variable name) that the resync happens every 30 runs, not 30 sec?